### PR TITLE
Skip generated remote share if exact user match found

### LINKED ISSUE #26647: Fed share appear in share autocomplete when searching for same email address
### Steps

1. Login as admin
1. Setup OC with a user "user@example.com" where "example.com" is the host of the current server
1. Create a folder "test"
1. Share "test" with "user@example.com"

### Expected result
Only one result "user@example.com"

### Actual result
Two results: "user@example.com" and "user@example.com (remote)"

Suggested quickfix: if there are two results with the same user id then remove the remote one, because the domain name matches the current server.

The fix could either be done on the JS side or better in the federated share provider, or in the code that gathers the result.

@cdamken

### PR DESCRIPTION
**Pull request by PVince81**
## Description

Whenever user ids are in the format "user@server", entering these in the share
dialog would also generate a remote share entry. Selecting the remote
entry would usually create a federated share to the same server.

In the case a user enters the full id address as share recipient it
is most likely that the intention was to share with the local user.

This fix removes the automatically generated remote id from the sharees
list when such situation occurs to avoid selecting the wrong entry by
mistake.

## Related Issue
Fixes https://github.com/owncloud/core/issues/26647

## Motivation and Context
Improve UX by preventing users to select the wrong share entry.

## How Has This Been Tested?
Manual testing, see steps in ticket + unit tests.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@DeepDiver1975 @jvillafanez @cdamken

### CODE REVIEW

**Comment by PVince81** [comment]
Please review @jvillafanez @VicDeo @DeepDiver1975 @butonic

**Comment by jvillafanez** [comment]
"user@server" could be a valid remote....

Isn't appending "(remote)" in the remote entry enough to distinguish the entries?

**Comment by PVince81** [comment]
> Isn't appending "(remote)" in the remote entry enough to distinguish the entries?

I was told that people selected the "(remote)" entry because they thought "oh yes, that guy is in another remote office" ...

In general if `user@server` is a valid remote AND `user@server` is used as account, it is very unlikely that `user@server (remote)` is about a separate server. It is very likely that it's exactly the same server, in which case it would be a local share. An admin wouldn't use this user id naming scheme if it wasn't the case I think.

**Comment by jvillafanez** [comment]
I could have a server in oc.com (for example), in which I can access though email. In that server I can have some guests (for example, j1@guest.com and j2@guest.com) who can also access to their respective accounts in my server.
Those guys could have their own ownCloud server in guest.com, and could link to their own files from my server.

If this situation is allowed (I don't think there is any reason to don't do it), if the j1@guest.com user wants to share a file to any user in the guest.com server, he won't be able to do it with this PR.

That's why I'd rather change the wording to make things clearer and keep the current behaviour

**Comment by PVince81** [comment]
Hmm, but if the OC server is oc.com, then the federated share id is then "j1@guest.com@oc.com" and "j2@guest.com@oc.com" and not the short email address. Because whatever you input the part behind the last @ is the actual target server. So if you only type "j1@guest.com" it would try to connect to "guest.com" which is no valid OC server.

@jvillafanez does that make sense ?

**Comment by jvillafanez** [comment]
But if there is a valid ownCloud in "guest.com", and there is a valid "j1" user there.... you won't be able to share anything with that user in that server.

**Comment by PVince81** [comment]
Hmmm... indeed...

@cdamken do you have a better suggestion maybe ? Change the wording of "(remote)" instead of removing that entry ? See previous comments, we can't remove it

**Comment by cdamken** [comment]
@PVince81 @jvillafanez 

The main problem here is if the owncloud is connected to the LDAP and the emails are the same as the owncloud (for example user carlos in damken,com cloud and the federated user will be carlos@damken.com) 
it appears 2 users:
```
carlos@damken.com 
carlos @damken.com  (remote)
```
The users sometimes select the ` remote`  (probably the logic is "the guy is not here in my office")

That generates a Federated share that is connected to the same server.

IMO a check if the remote request lands in the same server then should be created as local share.

**Comment by PVince81** [comment]
Checking if the remote lands on the same server is a much bigger issue: https://github.com/owncloud/core/issues/9090

**Comment by PVince81** [comment]
even if we do so the autocomplete would still display two entries.

Okay, @cdamken had another idea where we simply check if the domain part of the remote matches any of the trusted domains from config.php. If it does, we discard the entry.

@jvillafanez if you don't object to the latter I'll try that approach instead.

**Comment by cdamken** [comment]
> Okay, @cdamken had another idea where we simply check if the domain part of the remote matches any of the trusted domains from config.php. If it does, we discard the entry.

👍

**Comment by jvillafanez** [comment]
> Okay, @cdamken had another idea where we simply check if the domain part of the remote matches any of the trusted domains from config.php. If it does, we discard the entry.

I don't think it will work. The "damken.com" domain might not be in the config.php file. It's still risky.

**Comment by PVince81** [comment]
If "damken.com" is not in the trusted domains then the domain will not be accessible at all for users. So it has to be there.

**Comment by jvillafanez** [comment]
hmmm... basically, if the domain isn't in the config.php, no external sharing should be allowed. Kind of a whitelist, which makes me wonder if a sharing domain blacklist would be better. Maybe we should ask for opinions.

Anyway, this should apply for in and out links, I mean the remotes that are sent to the "forbidden" server as well as the remotes the "forbidden" server send to us, otherwise it would be weird.

**Comment by PVince81** [comment]
> if the domain isn't in the config.php, no external sharing should be allowed

Not at all. External sharing has nothing to do with trusted domains.
In the current implementation on master and stable branches external sharing is always allowed unless the admin disabled federated sharing in the admin page.

The trusted domain list in config.php is about web UI and Webdav clients. So if I have a domain "example.org" and don't add it to the list, then connecting to the web UI through "example.org" will be denied. If you add it to the list it will be allowed.

Regarding this PR, I would be only using the trusted domain list because this tells us what domain the current server has. Another alternative would be to use "overwrite.cli.url" or the current domain as displayed in the current browser page. And if the found domain matches the one from the result set, we discard the "(remote)" entry that has this domain because the server is local.

Hope this makes sense.

**Comment by jvillafanez** [comment]
Ok, make sense

**Comment by PVince81** [comment]
Added trusted_domain check.

But one unit test still fails and those unreadable huge data provider arrays make me go crazy...

**Comment by PVince81** [comment]
I failed to force push... Will look into the failing test first.

**Comment by PVince81** [comment]
Pushed now.

This should cover all discussed cases.

A) Given an ownCloud instance on domain "trusted.domain.tld" and two users "user1@trusted.domain.tld" and "user2@trusted.domain.tld" and logging in as the first user and sharing with the second one:
- typing "user2@trusted.domain.tld" will only show a single entry, the local one because it matches the domain
- typing "u@trusted.domain.tld" will show only the remote entry as there is no matching local user

B) Given an ownCloud instance on domain "trusted.domain.tld" any domain and two users "user1@guest.com" and "user2@guest.com"
- typing "user2@guest.com" will display both the local user and the remote one because it's not the same domain and user2 might have an own ownCloud instance on the domain guest.com.

Please review and test @cdamken @jvillafanez @davitol

File: apps/files_sharing/lib/Controller/ShareesController.php [hunk 1]
```diff
@@ -334,7 +334,14 @@ protected function getRemote($search) {
 			$this->result['remotes'] = [];
 		}
 
-		if (!$foundRemoteById && substr_count($search, '@') >= 1 && $this->offset === 0) {
+		if (!$foundRemoteById && substr_count($search, '@') >= 1 && $this->offset === 0
```
Comment (jvillafanez) [inline apps/files_sharing/lib/Controller/ShareesController.php]: We could move this condition to another function to improve readability.

File: apps/files_sharing/lib/Controller/ShareesController.php [hunk 2]
```diff
@@ -554,4 +561,27 @@ protected function getPaginationLink($page, array $params) {
...
+			// not a proper email-like format with domain name
+			return false;
+		}
+		$parts = explode('@', $target);
+		if (count($parts) === 1) {
+			// no "@" sign
+			return false;
+		}
+		$domainName = strtolower($parts[count($parts) - 1]);
+		$trustedDomains = $this->config->getSystemValue('trusted_domains', []);
```
Comment (jvillafanez) [inline apps/files_sharing/lib/Controller/ShareesController.php]: Maybe we should convert to lowercase the trusted domains, or consider an exact match

File: apps/files_sharing/lib/Controller/ShareesController.php [hunk 3]
```diff
@@ -554,4 +561,27 @@ protected function getPaginationLink($page, array $params) {
...
+
+		return in_array($domainName, $trustedDomains);
```
Comment (jvillafanez) [inline apps/files_sharing/lib/Controller/ShareesController.php]: I'd rather use a strict comparison: `in_array($domainName, $trustedDomains, true)`

File: apps/files_sharing/lib/MountProvider.php [hunk 4]
```diff
@@ -123,10 +123,12 @@ private function groupShares(array $shares) {
 		// sort by stime, the super share will be based on the least recent share
 		foreach ($tmp as &$tmp2) {
```
Comment (jvillafanez) [inline apps/files_sharing/lib/MountProvider.php]: We should open a ticket to review this piece of code. I think this is quite dangerous and there could be issues hiding here.
This is out of scope for this PR, but quite important to tackle it soon.

File: apps/files_sharing/lib/Controller/ShareesController.php (hunk 1 above)
Comment (PVince81) [inline apps/files_sharing/lib/Controller/ShareesController.php]: not now...

File: apps/files_sharing/lib/Controller/ShareesController.php (hunk 2 above)
Comment (PVince81) [inline apps/files_sharing/lib/Controller/ShareesController.php]: Let's do the exact match route then

File: apps/files_sharing/lib/MountProvider.php (hunk 4 above)
Comment (PVince81) [inline apps/files_sharing/lib/MountProvider.php]: uh oh... I didn't mean to check in this piece here... will remove

**Comment by PVince81** [comment]
@jvillafanez adjusted, please recheck. Thanks.

**Comment by jvillafanez** [comment]
👍

**Comment by davitol** [comment]
A) Given an ownCloud instance on domain "trusted.domain.tld" and two users "user1@trusted.domain.tld" and "user2@trusted.domain.tld" and logging in as the first user and sharing with the second one:

    typing "user2@trusted.domain.tld" will only show a single entry, the local one because it matches the domain ✅ 
    typing "u@trusted.domain.tld" will show only the remote entry as there is no matching local user ✅ 

B) Given an ownCloud instance on domain "trusted.domain.tld" any domain and two users "user1@guest.com" and "user2@guest.com"

    typing "user2@guest.com" will display both the local user and the remote one because it's not the same domain and user2 might have an own ownCloud instance on the domain guest.com. ✅ 


So WFM :+1:

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/27490
stable9: https://github.com/owncloud/core/pull/27491

**Pull request merged**